File: frontend/src/api/modules/topics.ts
import { request } from '@/utils/request';
import type {
  TopicRequest,
  TopicResponse,
  TopicFeasibilityRequest,
  TopicFeasibilityResponse,
  TopicRefinementRequest,
  TopicRefinementResponse,
  UserTopicsResponse
} from '@/types/topics';
import type { StreamEventHandler } from '@/types/stream';

/**
 * 推荐论文主题
 * @param data 主题请求参数
 */
export function recommendTopics(data: TopicRequest) {
  // 添加时间戳参数，避免缓存
  const timestamp = new Date().getTime();
  return request<TopicResponse[]>({
    url: `/topics/recommend?_t=${timestamp}`,
    method: 'post',
    data,
    headers: {
      'Cache-Control': 'no-cache',
      'Pragma': 'no-cache'
    }
  }).then(response => {
    // 打印响应数据以进行调试
    console.log('主题推荐原始响应数据:', response);

    // 检查响应数据是否有效
    if (!response) {
      console.error('响应数据为空');
      return [];
    }

    // 确保响应数据是数组
    const topics = Array.isArray(response) ? response : [response];
    console.log('处理后的主题数据:', topics);
    console.log('主题数量:', topics.length);

    // 验证每个主题对象是否完整
    const validatedTopics = topics.map(topic => {
      // 确保每个主题都有必要的字段
      return {
        title: topic.title || '未命名主题',
        research_question: topic.research_question || '无研究问题',
        feasibility: topic.feasibility || '未评估',
        innovation: topic.innovation || '无创新点',
        methodology: topic.methodology || '无研究方法',
        resources: topic.resources || '无资源需求',
        expected_outcomes: topic.expected_outcomes || '无预期成果',
        keywords: Array.isArray(topic.keywords) ? topic.keywords : []
      } as TopicResponse;
    });

    console.log('验证后的主题数据:', validatedTopics);
    return validatedTopics;
  }).catch(error => {
    console.error('获取主题推荐失败:', error);
    throw error;
  });
}

/**
 * 分析主题可行性
 * @param data 可行性分析请求参数
 */
export function analyzeTopicFeasibility(data: TopicFeasibilityRequest) {
  return request<TopicFeasibilityResponse>({
    url: '/topics/analyze',
    method: 'post',
    data
  });
}

/**
 * 优化论文主题
 * @param data 主题优化请求参数
 */
export function refineTopic(data: TopicRefinementRequest) {
  return request<TopicRefinementResponse>({
    url: '/topics/refine',
    method: 'post',
    data
  });
}

/**
 * 流式推荐论文主题
 * @param data 主题请求参数
 * @param eventHandlers 事件处理函数
 */
/**
 * 获取用户主题列表
 */
export function getUserTopics(skip: number = 0, limit: number = 100) {
  return request<UserTopicsResponse>({
    url: `/topics?skip=${skip}&limit=${limit}`,
    method: 'get'
  });
}

/**
 * 根据ID获取主题
 * @param id 主题ID
 */
export function getTopicById(id: number) {
  return request<TopicResponse>({
    url: `/topics/${id}`,
    method: 'get'
  });
}

export function recommendTopicsStream(data: TopicRequest, eventHandlers: StreamEventHandler) {
  // 添加时间戳参数，避免缓存
  const timestamp = new Date().getTime();
  const baseURL = import.meta.env.VITE_APP_API_BASE_URL || 'http://localhost:8000/api/v1';

  // 将请求参数编码到URL中
  const params = new URLSearchParams();
  params.append('user_interests', data.user_interests);
  params.append('academic_field', data.academic_field);
  if (data.academic_level) params.append('academic_level', data.academic_level);
  if (data.topic_count) params.append('topic_count', data.topic_count.toString());
  params.append('_t', timestamp.toString());

  // 构建URL
  const url = `${baseURL}/topics/recommend/stream?${params.toString()}`;

  console.log('创建SSE连接:', url);

  // 获取认证令牌
  const token = localStorage.getItem('token');

  if (!token) {
    console.error('未找到认证令牌，无法创建SSE连接');
    if (eventHandlers.onError) {
      eventHandlers.onError('未登录或会话已过期，请重新登录');
    }
    return () => {};
  }

  // 构建EventSource URL，包含认证令牌
  const eventSourceUrl = `${url}&token=${encodeURIComponent(token)}`;

  console.log('创建SSE连接，带有认证信息:', eventSourceUrl);

  // 创建EventSource对象
  // 注意：EventSource的withCredentials选项不会发送Authorization头，所以我们在URL中包含了token
  const eventSource = new EventSource(eventSourceUrl, { withCredentials: true });

  console.log('SSE连接创建成功，带有认证信息:', !!token);

  // 处理连接打开
  eventSource.onopen = (event) => {
    console.log('SSE连接已打开:', event);
    if (eventHandlers.onStatus) {
      eventHandlers.onStatus('连接已建立，等待数据...');
    }
  };

  // 处理消息事件
  eventSource.onmessage = (event) => {
    try {
      console.log('收到原始SSE数据:', event.data);
      const data = JSON.parse(event.data);
      console.log('解析后的流式数据:', data);

      // 根据消息类型调用相应的处理函数
      switch (data.type) {
        case 'status':
          if (eventHandlers.onStatus) {
            eventHandlers.onStatus(data.message);
          }
          break;
        case 'interest_analysis':
          if (eventHandlers.onInterestAnalysis) {
            eventHandlers.onInterestAnalysis(data.data);
          }
          break;
        case 'topic':
          if (eventHandlers.onTopic) {
            // 验证主题对象
            const topic = {
              title: data.data.title || '未命名主题',
              research_question: data.data.research_question || '无研究问题',
              feasibility: data.data.feasibility || '未评估',
              innovation: data.data.innovation || '无创新点',
              methodology: data.data.methodology || '无研究方法',
              resources: data.data.resources || '无资源需求',
              expected_outcomes: data.data.expected_outcomes || '无预期成果',
              keywords: Array.isArray(data.data.keywords) ? data.data.keywords : []
            } as TopicResponse;

            eventHandlers.onTopic(topic);
          }
          break;
        case 'complete':
          if (eventHandlers.onComplete) {
            eventHandlers.onComplete(data.message);
          }
          // 关闭EventSource
          eventSource.close();
          break;
        case 'error':
          if (eventHandlers.onError) {
            eventHandlers.onError(data.message);
          }
          // 关闭EventSource
          eventSource.close();
          break;
        default:
          console.warn('未知的消息类型:', data.type);
      }
    } catch (error) {
      console.error('解析流式数据失败:', error, event.data);
      if (eventHandlers.onError) {
        eventHandlers.onError('解析数据失败');
      }
    }
  };

  // 处理错误
  eventSource.onerror = (error) => {
    console.error('流式连接错误:', error);
    if (eventHandlers.onError) {
      eventHandlers.onError('流式连接错误，请检查网络或登录状态');
    }
    // 关闭EventSource
    eventSource.close();
  };

  // 返回关闭EventSource的函数
  return () => {
    console.log('关闭SSE连接');
    eventSource.close();
  };
}

File: backend/app/api/v1/endpoints/topics.py
from fastapi import APIRouter, Depends, HTTPException, Request, status
from fastapi.responses import StreamingResponse
from typing import List, Dict, Any, AsyncGenerator, Optional
from sqlalchemy.orm import Session
import json
import asyncio

from app.schemas.topic import (
    TopicRequest,
    TopicResponse,
    TopicFeasibilityRequest,
    TopicFeasibilityResponse,
    TopicRefinementRequest,
    TopicRefinementResponse
)
from app.services.topic_service import TopicService
from app.api.deps import get_topic_service, get_db, get_current_active_user, get_current_active_superuser, optional_oauth2_scheme
from app.models.user import User
from app.models.topic import Topic

router = APIRouter()

@router.post("/recommend", response_model=List[TopicResponse])
async def recommend_topics(
    request: TopicRequest,
    topic_service: TopicService = Depends(get_topic_service),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """推荐论文主题"""
    try:
        # 使用服务生成推荐主题
        topics = await topic_service.recommend_topics(
            user_interests=request.user_interests,
            academic_field=request.academic_field,
            academic_level=request.academic_level,
            topic_count=request.topic_count,
            interest_analysis=request.interest_analysis
        )

        # 将推荐的主题保存到数据库
        for topic_data in topics:
            # 检查topic_data是否为字典类型
            if isinstance(topic_data, dict):
                db_topic = Topic(
                    user_id=current_user.id,
                    title=topic_data.get('title', ''),
                    research_question=topic_data.get('research_question', ''),
                    feasibility=topic_data.get('feasibility', ''),
                    innovation=topic_data.get('innovation', ''),
                    methodology=topic_data.get('methodology', ''),
                    resources=topic_data.get('resources', ''),
                    expected_outcomes=topic_data.get('expected_outcomes', ''),
                    academic_field=request.academic_field,
                    academic_level=request.academic_level,
                    keywords=topic_data.get('keywords', [])
                )
            else:
                # 如果不是字典，尝试将其作为对象处理
                try:
                    db_topic = Topic(
                        user_id=current_user.id,
                        title=topic_data.title,
                        academic_field=request.academic_field,
                        academic_level=request.academic_level,
                        keywords=topic_data.keywords
                    )
                except Exception as e:
                    # 如果出错，记录错误并跳过这个主题
                    import logging
                    logging.error(f"处理主题数据失败: {str(e)}, 数据: {topic_data}")
                    continue

            db.add(db_topic)

        db.commit()
        return topics
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail=f"推荐主题失败: {str(e)}")

@router.post("/analyze", response_model=TopicFeasibilityResponse)
async def analyze_topic_feasibility(
    request: TopicFeasibilityRequest,
    topic_service: TopicService = Depends(get_topic_service),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """分析主题可行性"""
    try:
        # 分析主题可行性
        analysis = await topic_service.analyze_topic_feasibility(
            topic=request.topic,
            academic_field=request.academic_field,
            academic_level=request.academic_level
        )

        # 查找或创建主题记录
        db_topic = db.query(Topic).filter(
            Topic.user_id == current_user.id,
            Topic.title == request.topic
        ).first()

        if not db_topic:
            db_topic = Topic(
                user_id=current_user.id,
                title=request.topic,
                academic_field=request.academic_field,
                academic_level=request.academic_level
            )
            db.add(db_topic)

        # 更新可行性分析结果
        db_topic.feasibility = analysis.feasibility
        db_topic.innovation = analysis.innovation
        db_topic.methodology = analysis.methodology
        db_topic.resources = analysis.resources
        db_topic.expected_outcomes = analysis.expected_outcomes

        db.commit()
        return analysis
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail=f"分析主题可行性失败: {str(e)}")

@router.post("/refine", response_model=TopicRefinementResponse)
async def refine_topic(
    request: TopicRefinementRequest,
    topic_service: TopicService = Depends(get_topic_service),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """优化论文主题"""
    try:
        # 优化主题
        refined_topic = await topic_service.refine_topic(
            topic=request.topic,
            feedback=request.feedback,
            academic_field=request.academic_field,
            academic_level=request.academic_level
        )

        # 将优化后的主题保存到数据库
        db_topic = Topic(
            user_id=current_user.id,
            title=refined_topic.refined_title,
            research_question=refined_topic.research_question,
            academic_field=request.academic_field,
            academic_level=request.academic_level,
            keywords=refined_topic.keywords
        )
        db.add(db_topic)
        db.commit()

        return refined_topic
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail=f"优化主题失败: {str(e)}")

@router.get("/", response_model=List[dict])
async def get_user_topics(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
    skip: int = 0,
    limit: int = 100
):
    """获取用户的所有主题"""
    try:
        # 添加日志
        import logging
        logger = logging.getLogger("app")
        logger.info(f"用户 {current_user.username} (ID: {current_user.id}) 获取主题列表")

        topics = db.query(Topic).filter(Topic.user_id == current_user.id).offset(skip).limit(limit).all()
        result = [{
            "id": topic.id,
            "title": topic.title,
            "academic_field": topic.academic_field,
            "academic_level": topic.academic_level,
            "research_question": topic.research_question,
            "feasibility": topic.feasibility,
            "keywords": topic.keywords,
            "created_at": topic.created_at
        } for topic in topics]

        logger.info(f"找到 {len(result)} 条主题记录")
        return result
    except Exception as e:
        import logging
        logger = logging.getLogger("app")
        logger.error(f"获取主题列表失败: {str(e)}")
        raise HTTPException(status_code=500, detail=f"获取主题列表失败: {str(e)}")

@router.get("/admin/all", response_model=List[dict])
async def get_all_topics(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_superuser),
    skip: int = 0,
    limit: int = 100,
    user_id: Optional[int] = None
):
    """获取所有用户的主题（仅超级管理员可用）"""
    try:
        # 添加日志
        import logging
        logger = logging.getLogger("app")
        logger.info(f"超级管理员 {current_user.username} (ID: {current_user.id}) 获取所有主题列表")

        # 构建查询
        query = db.query(Topic)
        if user_id:
            query = query.filter(Topic.user_id == user_id)

        # 执行查询
        topics = query.order_by(Topic.created_at.desc()).offset(skip).limit(limit).all()

        # 获取用户信息
        user_ids = set(topic.user_id for topic in topics)
        users = {user.id: user for user in db.query(User).filter(User.id.in_(user_ids)).all()}

        result = [{
            "id": topic.id,
            "user_id": topic.user_id,
            "username": users.get(topic.user_id).username if topic.user_id in users else "Unknown",
            "title": topic.title,
            "academic_field": topic.academic_field,
            "academic_level": topic.academic_level,
            "research_question": topic.research_question,
            "feasibility": topic.feasibility,
            "keywords": topic.keywords,
            "created_at": topic.created_at,
            "updated_at": topic.updated_at
        } for topic in topics]

        logger.info(f"找到 {len(result)} 条主题记录")
        return result
    except Exception as e:
        import logging
        logger = logging.getLogger("app")
        logger.error(f"获取所有主题列表失败: {str(e)}")
        raise HTTPException(status_code=500, detail=f"获取所有主题列表失败: {str(e)}")

@router.get("/{topic_id}", response_model=dict)
async def get_topic_by_id(
    topic_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """根据ID获取主题"""
    try:
        topic = db.query(Topic).filter(Topic.id == topic_id, Topic.user_id == current_user.id).first()
        if not topic:
            raise HTTPException(status_code=404, detail="主题不存在")

        return {
            "id": topic.id,
            "title": topic.title,
            "academic_field": topic.academic_field,
            "academic_level": topic.academic_level,
            "research_question": topic.research_question,
            "feasibility": topic.feasibility,
            "innovation": topic.innovation,
            "methodology": topic.methodology,
            "resources": topic.resources,
            "expected_outcomes": topic.expected_outcomes,
            "keywords": topic.keywords,
            "created_at": topic.created_at,
            "updated_at": topic.updated_at
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"获取主题失败: {str(e)}")

@router.get("/recommend/stream")
async def recommend_topics_stream(
    request: Request,
    user_interests: str,
    academic_field: str,
    academic_level: str = "undergraduate",
    topic_count: int = 3,
    topic_service: TopicService = Depends(get_topic_service),
    db: Session = Depends(get_db)
):
    """流式推荐论文主题"""
    # 添加调试日志
    import logging
    from app.core.security import decode_access_token, SECRET_KEY, ALGORITHM
    from jose import jwt, JWTError
    logger = logging.getLogger("app")
    logger.info(f"SSE请求参数: user_interests={user_interests}, academic_field={academic_field}, academic_level={academic_level}, topic_count={topic_count}")

    # 从URL参数获取token
    token = request.query_params.get("token")
    if not token:
        logger.warning("URL参数中没有token")
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="未提供认证令牌",
            headers={"WWW-Authenticate": "Bearer"},
        )

    logger.info(f"从URL参数获取到token: {token[:10]}...")

    # 验证token
    credentials_exception = HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="无法验证凭据",
        headers={"WWW-Authenticate": "Bearer"},
    )

    try:
        payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
        username: str = payload.get("sub")
        if username is None:
            logger.warning("令牌中没有用户名")
            raise credentials_exception
        logger.info(f"令牌验证成功，用户名: {username}")

        # 获取用户信息
        current_user = db.query(User).filter(User.username == username).first()
        if current_user is None:
            raise credentials_exception

        user_id = current_user.id
        username = current_user.username
        user_info = f"用户: {username} (ID: {user_id})"
    except JWTError as e:
        logger.error(f"令牌验证失败: {str(e)}")
        raise credentials_exception

    logger.info(f"SSE请求用户: {user_info}")

    # 创建一个列表来存储生成的主题，以便在流式生成完成后保存到数据库
    generated_topics = []

    async def generate_stream() -> AsyncGenerator[bytes, None]:
        try:
            # 第一步：分析用户兴趣
            status_msg = json.dumps({"type": "status", "message": "正在分析用户兴趣..."})
            yield f"data: {status_msg}\n\n".encode('utf-8')

            # 异步分析用户兴趣
            interest_analysis = await topic_service.analyze_interests_streaming(
                user_interests=user_interests,
                academic_field=academic_field,
                academic_level=academic_level
            )

            # 将分析结果发送给前端
            analysis_msg = json.dumps({"type": "interest_analysis", "data": interest_analysis})
            yield f"data: {analysis_msg}\n\n".encode('utf-8')

            # 第二步：生成主题推荐
            recommend_msg = json.dumps({"type": "status", "message": "正在生成主题推荐..."})
            yield f"data: {recommend_msg}\n\n".encode('utf-8')

            # 流式生成主题
            async for topic in topic_service.recommend_topics_streaming(
                user_interests=user_interests,
                academic_field=academic_field,
                academic_level=academic_level,
                topic_count=topic_count,
                interest_analysis=interest_analysis
            ):
                # 将生成的主题添加到列表中，以便后续保存到数据库
                generated_topics.append(topic)

                # 将每个生成的主题发送给前端
                topic_msg = json.dumps({"type": "topic", "data": topic})
                yield f"data: {topic_msg}\n\n".encode('utf-8')

                # 模拟一些延迟，使前端可以看到渐进式生成
                await asyncio.sleep(0.5)

            # 将生成的主题保存到数据库
            try:
                logger.info(f"将 {len(generated_topics)} 个主题保存到数据库")
                for topic_data in generated_topics:
                    db_topic = Topic(
                        user_id=user_id,
                        title=topic_data.get('title', ''),
                        research_question=topic_data.get('research_question', ''),
                        feasibility=topic_data.get('feasibility', ''),
                        innovation=topic_data.get('innovation', ''),
                        methodology=topic_data.get('methodology', ''),
                        resources=topic_data.get('resources', ''),
                        expected_outcomes=topic_data.get('expected_outcomes', ''),
                        academic_field=academic_field,
                        academic_level=academic_level,
                        keywords=topic_data.get('keywords', [])
                    )
                    db.add(db_topic)

                db.commit()
                logger.info("主题成功保存到数据库")
            except Exception as e:
                db.rollback()
                logger.error(f"保存主题到数据库失败: {str(e)}")

            # 完成信号
            complete_msg = json.dumps({"type": "complete", "message": "主题生成完成"})
            yield f"data: {complete_msg}\n\n".encode('utf-8')

        except Exception as e:
            # 发送错误信息
            error_msg = json.dumps({"type": "error", "message": f"生成主题失败: {str(e)}"})
            yield f"data: {error_msg}\n\n".encode('utf-8')

    # 创建一个异步生成器包装器，在生成完成后关闭数据库连接
    async def wrapped_generator():
        try:
            async for chunk in generate_stream():
                yield chunk
        finally:
            # 无论生成是否成功或失败，都关闭数据库连接
            db.close()
            logger.info("数据库连接已关闭")

    # 创建流式响应
    return StreamingResponse(
        wrapped_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no"
        }
    )
